### ISSUE TITLE
Release?

### ISSUE THREAD

**Issue by jgarte** [Hexstream/enhanced-eval-when#1]
Hi, any plans for a release now that tests are added?

I'm packaging all your libraries for GNU/Guix ;()

https://guix.gnu.org/

**Comment by Hexstream** [comment]
> I'm packaging all your libraries for GNU/Guix ;()

**_WOW!!!_ Thank you!!** Honestly, I think this is the most attention my libraries have ever gotten so far... 😅

(**edit:** On second thought, Loopless had gotten somewhat a lot of attention back in the day, before I killed the project prematurely like a dumbass...)

> any plans for a release now that tests are added?

Honestly I wasn't really planning to resume development of my libraries until 2024,
since I'm planning to release [my most important project ever](https://roadmap.hexstreamsoft.com/far/#public-domain-spiritual-successor-to-the-clhs) in 2023.

That said, in this case, [turning this library green](https://www.hexstreamsoft.com/libraries/) is super low-hanging fruit, so I guess I might as well release a new version soon!

I'll keep you posted.

**Comment by jgarte** [comment]
Cool, keep me posted. 

Here's the patches I sent: 

https://issues.guix.gnu.org/57654

**Comment by Hexstream** [comment]
Alright!

Sorry about the delay, I needed to make a trivial update to all my 35 ready-to-use libraries and thus took the opportunity to [make absolutely extensive improvements to my GNU screen config](https://github.com/Hexstream/status-quo.hexstream.expert/commit/ef998da63bac4d50b94e9a735c0d94e9a48dcc8f), notably allowing me to instantly open dozens of terminals to any predefined subset of my 96 [git repos](https://abc.hexstream.xyz/repositories/) (such as all my libraries or all my websites, or all 96 repos), and [easily broadcast commands to all open terminals](https://github.com/Hexstream/status-quo.hexstream.expert/commit/ef998da63bac4d50b94e9a735c0d94e9a48dcc8f#diff-37d19747af6e8b2ba73c19e8703314f464d45930c89e75500a382b8a95c5d161R15-R17). **_This is simply life-changing!_**

Now, I am planning to **_probably_** make trivial updates and releases for the following [libraries](https://www.hexstreamsoft.com/libraries/) (as in, picking the super low-hanging fruit) in 2022, all at once:

- [enhanced-eval-when](https://www.hexstreamsoft.com/libraries/enhanced-eval-when/)
- [enhanced-multiple-value-bind](https://www.hexstreamsoft.com/libraries/enhanced-multiple-value-bind/)
- [macro-level](https://www.hexstreamsoft.com/libraries/macro-level/)
- [its](https://www.hexstreamsoft.com/libraries/its/)
- [parse-number-range](https://www.hexstreamsoft.com/libraries/parse-number-range/)
- [bubble-operator-upwards](https://www.hexstreamsoft.com/libraries/bubble-operator-upwards/)

After that, I'll probably do the major update and release for [cartesian-product-switch](https://www.hexstreamsoft.com/libraries/cartesian-product-switch/).

After that, probably no more library updates before 2024.

I hope this is to your satisfaction!

**Comment by jgarte** [comment]
That sounds awesome! Let me know and I'll update all the libs. I might need the GNU screen config also :)

Why GNU screen and not tmux? You just have more familiarity with screen?

**Comment by Hexstream** [comment]
> That sounds awesome! Let me know and I'll update all the libs.

Thank you!! That's very motivating!!

> I might need the GNU screen config also :)

[It's here](https://github.com/Hexstream/status-quo.hexstream.expert/tree/master/assets/tools/screen/config)!

I'll update [my usage notes](https://status-quo.hexstream.expert/tools/screen/) later.

> Why GNU screen and not tmux? You just have more familiarity with screen?

GNU screen is the traditional choice and what I started with, and there doesn't seem to be enough incentive to switch to tmux, especially since I don't use the layout features. And until recently, my usage of `screen` was very basic. Now it's quite a bit more advanced than it used to be, but I haven't yet run into significant limitations I couldn't work around.

Ideally, I'd rather use a `screen` replacement written in Common Lisp...

**Comment by jgarte** [comment]
> Ideally, I'd rather use a screen replacement written in Common Lisp...

That sounds like a nice [croatoan](https://github.com/McParen/croatoan) project.

**Comment by Hexstream** [comment]
Perhaps, but I have a lot of much higher priorities on [my roadmap](https://roadmap.hexstreamsoft.com/). ;)

**Comment by jgarte** [comment]
Yes, I'm looking forward to package those higher priorities for [GNU Guix](guix.gnu.org) ;()

**Comment by jgarte** [comment]
@Hexstream Is the "public domain spiritual successor to CLHS" open to public contribution? That sounds like a project that would benefit from crowd sourced knowledge and writing contributions.

For example, I started contributing to the common lisp cookbook here: https://github.com/LispCookbook/cl-cookbook/pull/458

That FFI knowledge came after I had a session with [phoe](https://phoe.github.io/). I'm looking to finish that PR soon.

**Comment by Hexstream** [comment]
> @Hexstream Is the "public domain spiritual successor to CLHS" open to public contribution? That sounds like a project that would benefit from crowd sourced knowledge and writing contributions.
> 
> For example, I started contributing to the common lisp cookbook here: [LispCookbook/cl-cookbook#458](https://github.com/LispCookbook/cl-cookbook/pull/458)

Yes, once the initial version is published, just like for all my other projects.

In fact, an option for extensive linking to the Cookbook would be highly desirable as an **_optional_** feature.
(I don't think that content would be integrated directly in my version since I want to keep everything public domain.)

I am planning to be VERY conservative about the _default_ presentation of [my CLHS replacement](https://roadmap.hexstreamsoft.com/far/#public-domain-spiritual-successor-to-the-clhs), it will just be a better version of the CLHS, period. After the initial release, I will likely add easy options to load additional helpful modules on request, of which extensive linking to the Cookbook would be one.

----------

> That FFI knowledge came after I had a session with [phoe](https://phoe.github.io/). I'm looking to finish that PR soon.

Uh-huh. [I am no fan of Scambird](https://hexstream.link/exposed/phoe).
He appears to have somehow reached the exact opposite conclusion,
but I certainly don't care to try to make sense of it.

(I no longer track his scams as of 25 may 2021.)

**Comment by jgarte** [comment]
>  He appears to have somehow reached the exact opposite conclusion,

Sorry to hear about that situation. I read your post. I hope things will work themselves out for the better in the future. It's such a small community and we need all the teamwork we can afford to make Lisp live on. As Alexander poetically describes it: 40 hard working ants. I'm very new to the Common Lisp community.

> I am planning to be VERY conservative about the default presentation of [my CLHS replacement]

Is there a repo for the CLHS replacement yet?

I'll keep you posted on the what I send to GNU Guix. I just sent this patch for `bubble-operator-upwards` which will probably get merged soon:

https://issues.guix.gnu.org/57917

**Comment by Hexstream** [comment]
> Sorry to hear about that situation. I read your post. I hope things will work themselves out for the better in the future. It's such a small community and we need all the teamwork we can afford to make Lisp live on. As Alexander poetically describes it: 40 hard working ants. I'm very new to the Common Lisp community.

Awesome! **_Welcome!!_**

I do expect that situation to naturally resolve itself over the years, actually...

> Is there a repo for the CLHS replacement yet?

No, these days I usually like to keep my repos private initially, and then rebase the entire history into one "initial commit" immediately before the first public release.

> I'll keep you posted on the what I send to GNU Guix. I just sent this patch for bubble-operator-upwards which will probably get merged soon:
>
> https://issues.guix.gnu.org/57917

Thank you, although I have no current plans to review that information, I'll just leave this in your capable hands! ;)

**Comment by Hexstream** [comment]
Sorry, unfortunately there have been some unexpected delays, and this is ongoing.

[I will update my libraries in 2024](https://roadmap.hexstreamsoft.com/#update-ready-to-use-libraries), as I had originally planned.

(It's also much better for efficiency and momentum to do this all in one shot rather than split it into two phases...)

I apologize for any inconvenience.

**Comment by jgarte** [comment]
> I am planning to be VERY conservative about the _default_ presentation of [my CLHS replacement](https://roadmap.hexstreamsoft.com/far/#public-domain-spiritual-successor-to-the-clhs), it will just be a better version of the CLHS, period. After the initial release, I will likely add easy options to load additional helpful modules on request, of 

WDYT of releasing your CLHS as open source or public domain/Creative Commons so I and others can help with developing it along with you?

**Comment by Hexstream** [comment]
I will do that, _once the initial version is done_.

I won't do that _before_ for a few different reasons, such as:

- I want all the glory for myself. ;P (And **I actually _need_ it**, for stupid political reasons.)
- I want to avoid any risk of my contribution being misrepresented as much smaller than it actually is.
- I simply DO NOT want to wait on I/O, which would mostly kill my productivity.
- I simply DO NOT want to manage people, which would mostly kill my productivity.
- Adding people to the project would probably just add overhead and slow me down.
- I want to do this project 100% like I want, without needing to consult or get "buy-in" from anyone.
- I will do this project mostly manually, which most people probably couldn't bear at all. (Although, I will automate some crucial parts with keyboard macros, thereby saving months of truly soul-crushing work.)
- I want to be able to make drastic changes without invalidating anyone else's work.
- I want to avoid the risk of someone forking the project and finishing it before I'm done.
- I have **_EXTENSIVE_** experience working this way, and using a completely different process on my most important project ever doesn't seem wise at all.
- I don't want (almost certainly premature and/or useless) criticism on my WIP while I'm working on the initial version.
- It's not even clear that anyone would want to help me with this project, and if they did, that their contribution would actually help me significantly, even before considering the drawbacks and overheads of their contributions. If there was very little demonstrated interest during development, this could furthermore be used to falsely represent my project as not that interesting or useful after all. And _literally_ don't even get me started on the fact that any actually useful contributors I might attract would almost certainly quickly get scammed and/or terrorismed away, probably before they even contact me.
- I would frankly have to consider the _very real and terrifying possibility_ that any previously unknown people _seemingly_ trying to help me could be sabotageurs. The consequences could easily be _truly devastating_ in any number of ways. **I simply cannot afford suffering [yet](https://hexstream.link/exposed/xach) [another](https://hexstream.link/exposed/phoe) high-profile and impactful betrayal** anytime soon.
- Probably a bunch of other good reasons, but I think this is enough for now.

I hope this is to your satisfaction. ;P

**Comment by Hexstream** [comment]
[I upgraded to tmux!](https://status-quo.hexstream.expert/#18_january_2023)

**Comment by Hexstream** [comment]
[Done!](https://www.hexstreamsoft.com/#6_october_2023) Thank you for the prompt.

**Comment by jgarte** [comment]
Looks like this version breaks the ecl build:

```
warning: collision encountered:
  /gnu/store/na1dpbbcxjaa3n8wkwrfpch476f90hlf-ld-wrapper-0/bin/ld
  /gnu/store/zh4x65snfis7svs6906gj1z8i7dx2j3m-binutils-2.38/bin/ld
warning: choosing /gnu/store/na1dpbbcxjaa3n8wkwrfpch476f90hlf-ld-wrapper-0/bin/ld
phase `configure' succeeded after 0.3 seconds
starting phase `patch-generated-file-shebangs'
phase `patch-generated-file-shebangs' succeeded after 0.1 seconds
starting phase `build'
Invoking ecl: "/gnu/store/gbahff6l8b6413vx52klkljr7ypmpz6c-ecl-23.9.9/bin/ecl" "--eval" "(require :asdf)" "--eval" "(asdf:initialize-source-registry (list :source-registry (list :tree (uiop:ensure-pathname \"/gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs\" :truenamize t :ensure-directory t)) :inherit-configuration))" "--eval" "(asdf:load-system \"canonicalized-initargs\")" "--eval" "(quit)" 
;;; Loading #P"/gnu/store/gbahff6l8b6413vx52klkljr7ypmpz6c-ecl-23.9.9/lib/ecl-23.9.9/asdf.fas"
;;; Warning: System definition file #P"/gnu/store/mbfg2h3x2k9agiybwzj882al4xcz9mhy-ecl-trivial-garbage-0.21/share/common-lisp/ecl/trivial-garbage/trivial-garbage.asd" contains definition for system "trivial-garbage-tests". Please only define "trivial-garbage" and secondary systems with a name starting with "trivial-garbage/" (e.g. "trivial-garbage/test") in that file.
;;; Warning: Class CLASS has been forward referenced.
;;;
;;; Compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/package.lisp.
;;; OPTIMIZE levels: Safety=2, Space=0, Speed=3, Debug=0
;;;
;;; Finished compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/package.lisp.
;;;
;;;
;;; Compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/shared.lisp.
;;; OPTIMIZE levels: Safety=2, Space=0, Speed=3, Debug=0
;;;
;;; Finished compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/shared.lisp.
;;;
;;;
;;; Compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/canonicalization-strategy.lisp.
;;; OPTIMIZE levels: Safety=2, Space=0, Speed=3, Debug=0
;;;
;;; Finished compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/canonicalization-strategy.lisp.
;;;
;;;
;;; Compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/canonicalization-class.lisp.
;;; OPTIMIZE levels: Safety=2, Space=0, Speed=3, Debug=0
;;;
;;; Finished compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/canonicalization-class.lisp.
;;;
;;;
;;; Compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/slot-initargs.lisp.
;;; OPTIMIZE levels: Safety=2, Space=0, Speed=3, Debug=0
;;;
;;; Finished compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/slot-initargs.lisp.
;;;
;;;
;;; Compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/class-initargs.lisp.
;;; OPTIMIZE levels: Safety=2, Space=0, Speed=3, Debug=0
;;;
;;; Finished compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/class-initargs.lisp.
;;;
;;;
;;; Compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/standard.lisp.
;;; OPTIMIZE levels: Safety=2, Space=0, Speed=3, Debug=0
;;;
;;; Finished compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/standard.lisp.
;;;
phase `build' succeeded after 5.3 seconds
starting phase `check'
Invoking ecl: "/gnu/store/gbahff6l8b6413vx52klkljr7ypmpz6c-ecl-23.9.9/bin/ecl" "--eval" "(require :asdf)" "--eval" "(asdf:initialize-source-registry (list :source-registry (list :tree (uiop:ensure-pathname \"/gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs\" :truenamize t :ensure-directory t)) :inherit-configuration))" "--eval" "(asdf:test-system \"canonicalized-initargs\")" "--eval" "(quit)" 
;;; Loading #P"/gnu/store/gbahff6l8b6413vx52klkljr7ypmpz6c-ecl-23.9.9/lib/ecl-23.9.9/asdf.fas"
;;; Warning: System definition file #P"/gnu/store/mbfg2h3x2k9agiybwzj882al4xcz9mhy-ecl-trivial-garbage-0.21/share/common-lisp/ecl/trivial-garbage/trivial-garbage.asd" contains definition for system "trivial-garbage-tests". Please only define "trivial-garbage" and secondary systems with a name starting with "trivial-garbage/" (e.g. "trivial-garbage/test") in that file.
;;; Warning: Class CLASS has been forward referenced.
;;;
;;; Compiling /gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs/tests/tests.lisp.
;;; OPTIMIZE levels: Safety=2, Space=0, Speed=3, Debug=0
;;;
;;; Error:
;;;   in file tests.lisp, position 507
;;;   at (DEFCLASS CANON-BOOLEAN ...)
;;;   * The macro form (DEFCLASS CANON-BOOLEAN NIL ((%BOOLEAN :INITARG :BOOLEAN :READER %BOOLEAN :CANONICALIZE (FUNCTION BOOLEAN)))) was not expanded successfully.
;;; Error detected:
;;; In function CAR, the value of the first argument is
An error occurred during initialization:
#<a FLOATING-POINT-INVALID-OPERATION 0x7ffff4b624c0>.
error: in phase 'check': uncaught exception:
%exception #<&invoke-error program: "/gnu/store/gbahff6l8b6413vx52klkljr7ypmpz6c-ecl-23.9.9/bin/ecl" arguments: ("--eval" "(require :asdf)" "--eval" "(asdf:initialize-source-registry (list :source-registry (list :tree (uiop:ensure-pathname \"/gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs\" :truenamize t :ensure-directory t)) :inherit-configuration))" "--eval" "(asdf:test-system \"canonicalized-initargs\")" "--eval" "(quit)") exit-status: 1 term-signal: #f stop-signal: #f> 
phase `check' failed after 1.1 seconds
command "/gnu/store/gbahff6l8b6413vx52klkljr7ypmpz6c-ecl-23.9.9/bin/ecl" "--eval" "(require :asdf)" "--eval" "(asdf:initialize-source-registry (list :source-registry (list :tree (uiop:ensure-pathname \"/gnu/store/bbxlz5ivjv5y95w5qpc2rhgl616r1slj-ecl-canonicalized-initargs-2.0/share/common-lisp/ecl/canonicalized-initargs\" :truenamize t :ensure-directory t)) :inherit-configuration))" "--eval" "(asdf:test-system \"canonicalized-initargs\")" "--eval" "(quit)" failed with status 1
builder for `/gnu/store/k6jbdxn4k4b7k92y0zjfv08knvgxclk0-ecl-canonicalized-initargs-2.0.drv' failed with exit code 1
build of /gnu/store/k6jbdxn4k4b7k92y0zjfv08knvgxclk0-ecl-canonicalized-initargs-2.0.drv failed
View build log at '/var/log/guix/drvs/k6/jbdxn4k4b7k92y0zjfv08knvgxclk0-ecl-canonicalized-initargs-2.0.drv.gz'.
cannot build derivation `/gnu/store/24sw54vr2y8lr5svqxi5szavnqvrfdmk-ecl-definitions-systems-2.0.1.drv': 1 dependencies couldn't be built
guix build: error: build of `/gnu/store/24sw54vr2y8lr5svqxi5szavnqvrfdmk-ecl-definitions-systems-2.0.1.drv' failed
```
